File: go/internal/cache/cache.go
package cache

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"log"
	"os"
	"path/filepath"
	"time"

	"codex-web-mcp/internal/config"
)

type Entry struct {
	Content   string `json:"content"`
	Timestamp string `json:"timestamp"`
}

func KeyHash(tool, payload string) string {
	h := sha256.Sum256([]byte(tool + "::" + payload))
	return hex.EncodeToString(h[:])[:32]
}

func filePath(tool, payload string) (string, error) {
	sub, err := config.CacheSubdir()
	if err != nil {
		return "", err
	}
	return filepath.Join(sub, fmt.Sprintf("%s-%s.json", tool, KeyHash(tool, payload))), nil
}

func Get(tool, payload string, ttlSeconds int) (string, bool) {
	p, err := filePath(tool, payload)
	if err != nil {
		return "", false
	}
	info, err := os.Stat(p)
	if err != nil {
		return "", false
	}
	if time.Since(info.ModTime()) > time.Duration(ttlSeconds)*time.Second {
		return "", false
	}
	data, err := os.ReadFile(p)
	if err != nil {
		return "", false
	}
	var e Entry
	if err := json.Unmarshal(data, &e); err != nil {
		return "", false
	}
	return e.Content, true
}

func Put(tool, payload, content string) {
	p, err := filePath(tool, payload)
	if err != nil {
		log.Printf("cache put: %v", err)
		return
	}
	e := Entry{Content: content, Timestamp: time.Now().UTC().Format(time.RFC3339)}
	data, err := json.Marshal(&e)
	if err != nil {
		log.Printf("cache marshal: %v", err)
		return
	}
	if err := os.WriteFile(p, data, 0o644); err != nil {
		log.Printf("cache write: %v", err)
		return
	}
}

File: node/src/cache.ts
import * as crypto from "node:crypto";
import * as fs from "node:fs";
import * as path from "node:path";

import { cacheSubdir } from "./config.js";

interface Entry {
  content: string;
  timestamp: string;
}

export function keyHash(tool: string, payload: string): string {
  return crypto.createHash("sha256").update(`${tool}::${payload}`).digest("hex").slice(0, 32);
}

function filePath(tool: string, payload: string): string {
  return path.join(cacheSubdir(), `${tool}-${keyHash(tool, payload)}.json`);
}

export function get(tool: string, payload: string, ttlSeconds: number): string | null {
  try {
    const p = filePath(tool, payload);
    const stat = fs.statSync(p);
    const ageMs = Date.now() - stat.mtimeMs;
    if (ageMs > ttlSeconds * 1000) return null;
    const data = fs.readFileSync(p, "utf8");
    const entry = JSON.parse(data) as Entry;
    return typeof entry.content === "string" ? entry.content : null;
  } catch {
    return null;
  }
}

export function put(tool: string, payload: string, content: string): void {
  try {
    const p = filePath(tool, payload);
    const entry: Entry = { content, timestamp: new Date().toISOString() };
    fs.writeFileSync(p, JSON.stringify(entry), "utf8");
  } catch (e) {
    process.stderr.write(`[codex-web-mcp] cache write failed: ${(e as Error).message}\n`);
  }
}
